woocommerce/woocommerce
# Replace @wordpress/components Button, Radio, RadioGroup with Ariakit Button

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/46062 https://github.com/woocommerce/woocommerce/issues/46208

This introduces the Ariakit `Button` component. We use it to replace our use of the `@wordpress/components` button, but also to replace `__experimentalRadioGroup` and `__experimentalRadio`.

I considered using Ariakit `RadioGroup` and `Radio` for these, here's a summary of my reasoning not to:

* They are used for one single feature in checkout, selecting between pickup and shipping, and their styling / functionality is more like 2 giant buttons that select between 2 values managed in state, not actual form data. For aria we can just apply roles of `radiogroup` and `radio`.

* We already have some existing state management of 2 simple values, if we used RadioGroup and Radio we'd need to introduce not only RadioGroup, Radio but also the store and state management for those components and we just don't need them. Furthermore this is the **only** instance where we use the old wordpress components for pseudo radio functionality, so introducing a whole new dependency here is just not worth it.

* Also if we did introduce Radio/RadioGroup styling them was going to be a lot more work, they are radio buttons by default, whereas with using buttons here we get to make very few changes to get button styles  because the wordpress component (which is now deprecated) was also using buttons under the hood.

#### Deprecation of showSpinner

In the past on WooCommerce we've tended to have API's like `showSpinner` for components that will trigger the rendering or not rendering of some child component like a spinner. I've deprecated this because I think it's always better to render something like that in `children` of the component. It's more flexible in general and especially with a primitive component it is better not to render any child content in the component by default because it's not typically needed. It also allows for better performance in future, if we don't bundle components together in this way, if an entrypoint uses button but never with a spinner, for example, it should not need a dependency on `Spinner`.

To achieve the same result you can render <Spinner> as a child when you have a loading state.

e.g.

```
<Button>
   {isLoading && <Spinner />}
</Button>
```

#### Removing support for icon

For the same reasoning as `showSpinner` we don't need to introduce opinionated ways that children are rendered in primitive components. One can always use composition in future, if rendering a button with an icon is needed (ie create a IconButton component). 

Furthermore, we never documented that the API of `@wordpress/components` button was the public API of our component, it just so happens in a few places we relied on the passing down of all props to use the icon prop, but there are other props like `iconPosition` and others that one could technically use in the current implementation. I think it's better to remove usage of any underlying props of wp button instead of trying to support backwards compat for them as we never advertised they were part of the API.

To achieve the same result of rendering an icon as before, you should render it as a child of the button as we've done in this PR.

e.g.

```
<Button>
  <Icon icon={ close } />
</Button>
```



### How to test the changes in this Pull Request:

* Mini cart close button (top right of open mini-cart)
* All buttons and link buttons in the cart and checkout
* Close button on the notice component
* Local pickup / Shipping "buttons" in the checkout form



### Changelog entry




<details>

#### Significance




#### Type




#### Message 
In blocks migrate `@wordpress/components` Button to Ariakit, replace  `__experimentalRadio/RadioGroup` with Ariakit Button.

#### Comment 

</details>
